### PR TITLE
[WIP] dotnet-installer uses Tls 1.2

### PR DESCRIPTION
**Pull request by bekir-ozturk**
Install script was failing since download links can no longer be accessed with Tls1.1.
Explicitly adding Tls1.2 to SecurityProtocol to fix the issue.

### CODE REVIEW

**Review by mmitche — APPROVED** [review]

**Comment by wli3** [comment]
We need to wait for the new scripts get deployed

**Comment by bozturkMSFT** [comment]
> We need to wait for the new scripts get deployed

What new scripts are we talking about here?

**Comment by mmitche** [comment]
New dotnet-install.ps1.  Should be good to go now.

**Comment by marcpopMSFT** [comment]
See mail thread before merging this.  It looks like adding 1.2 may exclude 1.3 support which will impact some customers.

**Comment by mmitche** [comment]
Yep, I'm pretty sure that's what's happening. The problem is that you can't add to the SystemDefault.

**Comment by reznet** [comment]
Did this get deployed somehow already?  Our automation, which downloads this script every time, started to fail with the error:

```
System.Management.Automation.CmdletInvocationException: Exception setting "SecurityProtocol": "The requested security protocol is not supported." ---> System.Management.Automation.SetValueInvocationException: Exception setting "SecurityProtocol": "The requested security protocol is not supported." ---> System.NotSupportedException: The requested security protocol is not supported.
```

but the problem went away.  We're going to move to run our own copy of the script, but would also recommend not deploying this change as is.

**Comment by nohwnd** [comment]
@reznet I had the same issue, if by a chance you are running against vs2017, it helped me to move to vs2019 image for our internal pipeline.

**Comment by mmitche** [comment]
> Did this get deployed somehow already? Our automation, which downloads this script every time, started to fail with the error:
> 
> ```
> System.Management.Automation.CmdletInvocationException: Exception setting "SecurityProtocol": "The requested security protocol is not supported." ---> System.Management.Automation.SetValueInvocationException: Exception setting "SecurityProtocol": "The requested security protocol is not supported." ---> System.NotSupportedException: The requested security protocol is not supported.
> ```
> 
> but the problem went away. We're going to move to run our own copy of the script, but would also recommend not deploying this change as is.

We reverted it. I'm not sure there is actually a way to enable this safely (at least in powershell) and it will need to be fixed on the caller side. It's possible that your systems do not have strong crypto enabled.

**Comment by nohwnd** [comment]
@mmitche Trying to set it from the highest version? Not sure which exception from the ones you listed would be caught, so I simply fallback on any exception. 


```powershell
try {
    [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::Tls13
} 
catch  { 
    try {
        [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::Tls12
    }
    catch {
        [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::SystemDefault
    }
}
```

**Comment by mmitche** [comment]
> @mmitche Trying to set it from the highest version? Not sure which exception from the ones you listed would be caught, so I simply fallback on any exception.
> 
> ```powershell
> try {
>     [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::Tls13
> } 
> catch  { 
>     try {
>         [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::Tls12
>     }
>     catch {
>         [System.Net.ServicePointManager]::SecurityProtocol += [System.Net.SecurityProtocolType]::SystemDefault
>     }
> }
> ```

I hadn't actually gotten it to fail while setting.  To any value, really. It only fails during download.

**Comment by nohwnd** [comment]
Right, that is how I see people testing if a given version of tls is supported on the system as well, by making a request with that setting. 

The only option would be to put the try catch around the download then. If you can repro locally then just check what is the exception that is thrown and specify it like this: 
```powershell
try {

} 
catch [System.NotSupportedException] {
}
```

**Comment by bozturkMSFT** [comment]
> System.NotSupportedException: The requested security protocol is not supported.

This exception is thrown if you try to set Tls1.2 flag when it is already set.
If you call the script repeatedly, first attempt would succeed, but second would fail.

I replaced first '+=' with '=' so that existing flags will be cleared.
Marking WIP so that we don't accidentally merge this.

File: scripts/obtain/dotnet-install.ps1 [hunk 1]
```diff
@@ -118,6 +118,11 @@ if ($SharedRuntime -and (-not $Runtime)) {
 $VersionRegEx="/\d+\.\d+[^/]+/"
 $OverrideNonVersionedFiles = !$SkipNonVersionedFiles
 
+# Default Tls protocol for .NET Framework 4.5 does not allow for Tls1.2.
+[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;
```
Comment (filipnavara) [inline scripts/obtain/dotnet-install.ps1:122]: Should not this be `+=`? This will override the system defaults even in case of some hypothetical Tls14.

**Comment by marcpopMSFT** [comment]
@bozturkMSFT did you find the right solution for TLS 1.2 and is this PR still needed?

**Comment by bozturkMSFT** [comment]
@grinrag is working on the TLS 1.2 issue now. I don't think this PR will be the solution. Closing.

**Comment by bozturkMSFT** [comment]
Yes, here it is https://github.com/dotnet/sdk/issues/11320

File: scripts/obtain/dotnet-install.ps1 (hunk 1 above)
Comment (bergmeister) [inline scripts/obtain/dotnet-install.ps1:122]: You shouldn't be changing all enabled protocols of the user here but rather `ADD` only the ones you need, like you do in line 124

Comment (bozturkMSFT) [reply scripts/obtain/dotnet-install.ps1:122]: `+=` isn't safe to use if `Tls12` flag is already set. I guess the correct way to fix this is to use "binary or" `-bor`.

**Pull request closed without merging**